avail/src/main/kotlin/avail/anvil/shortcuts/ModifierKey.kt

```diff
@@ -0,0 +1,183 @@
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ *  * Neither the name of the copyright holder nor the names of the contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package avail.anvil.shortcuts
+
+import java.awt.Toolkit
+import java.awt.event.InputEvent
+import javax.swing.KeyStroke
+
+/**
+ * Enumerates the [InputEvent] modifier keys.
+ *
+ * @author Richard Arriaga
+ *
+ * @property modifier
+ *   The modifier constant.
+ * @property lookupKey
+ *   The key used to look up this modifier.
+ */
+enum class ModifierKey constructor(
+	val modifier: Int,
+	val lookupKey: String)
+{
+	/** The [Shift Key][InputEvent.SHIFT_DOWN_MASK]. */
+	SHIFT_DOWN_MASK(InputEvent.SHIFT_DOWN_MASK, "SHIFT")
+	{
+		override val displayRepresentation: String = "⇧"
+	},
+
+	/** The [Control Key][InputEvent.CTRL_DOWN_MASK]. */
+	CTRL_DOWN_MASK(InputEvent.CTRL_DOWN_MASK, "CTRL")
+	{
+		override val displayRepresentation: String = "⌃"
+	},
+
+	/**
+	 * The [Meta Key][InputEvent.META_DOWN_MASK]. On Windows this is the Winodws
+	 * key; on Mac, the Command key.
+	 */
+	META_DOWN_MASK(InputEvent.META_DOWN_MASK, "META")
+	{
+		override val displayRepresentation: String get() =
+			if (System.getProperty("os.name").startsWith("Mac"))
+			{
+				"⌘"
+			}
+			else
+			{
+				"Windows"
+			}
+	},
+
+	/** The [Alt Key][InputEvent.ALT_DOWN_MASK]. */
+	ALT_DOWN_MASK(InputEvent.ALT_DOWN_MASK, "ALT")
+	{
+		override val displayRepresentation: String = "⌥"
+	},
+
+	/** The [Mouse Button 1][InputEvent.BUTTON1_DOWN_MASK]. */
+	BUTTON1_DOWN_MASK(InputEvent.BUTTON1_DOWN_MASK, "LEFT-CLICK"),
+
+	/** The [Mouse Button 2][InputEvent.BUTTON2_DOWN_MASK]. */
+	BUTTON2_DOWN_MASK(InputEvent.BUTTON2_DOWN_MASK, "RIGHT-CLICK"),
+
+	/** The [Mouse Button 3][InputEvent.BUTTON3_DOWN_MASK]. */
+	BUTTON3_DOWN_MASK(InputEvent.BUTTON3_DOWN_MASK, "MIDDLE-MOUSE-BUTTON");
+
+	/**
+	 * The platform-specific String that represents this key.
+	 */
+	open val displayRepresentation: String get () = lookupKey
```

Comment:
Actually this is where the pattern originated because of the `META` key which has a different representation based on the OS. I took a look at how IntelliJ represents this key in its key map preferences. To represent this key on Mac they use `⌘` but on Windows and Ubuntu they use `Windows`. So I did this so that I could specifically do the following.

```kotlin
META(InputEvent.META_DOWN_MASK, "META")
{
	override val displayRepresentation: String get() =
		if (System.getProperty("os.name").startsWith("Mac"))
		{
			"⌘"
		}
		else
		{
			"Windows"
		}
}
```

I just blindly copied this pattern elsewhere as I guess I had some expectation this might happen again but as you observed I didn't need to use this pattern elsewhere and so I have fixed it elsewhere.

I'm open to ideas on how to potentially address this more cleanly for `ModifierKey` given the `META` key situation.

Reply:
Heh, the `os.name` isn't going to change while the program's running, so you can make the if/else decision once, during enum initialization.  Try this (I haven't test-compiled it):
```
enum class ModifierKey constructor(
	val modifier: Int,
	val lookupKey: String,
	val displayRepresentation: String = lookupKey)
{
	...
	META(
		InputEvent.META_DOWN_MASK,
		"META",
		System.getProperty("os.name").let { os ->
			when
			{
				os.startsWith("Mac") -> "⌘"
				os.startsWith("Windows") -> "Windows"
				else -> "META"
			}
		})
	...
}
```